### PR TITLE
Bug 2032559: Block DualStack migration for unsupported cluster types

### PR DESCRIPTION
**Pull request by andreaskaris**
Only allow DualStack migration for BareMetal and None type clusters.

Additionally, update error messages for DualStack migrations.
Now that we are not hiding the existence of DualStack any more, give
clearer error messages for unsupported configurations.

### CODE REVIEW

File: pkg/network/ovn_kubernetes.go
Comment (flavio-fernandes) [inline pkg/network/ovn_kubernetes.go:271]: Maybe use Errorf instead of Infof ?

**Review by flavio-fernandes — COMMENTED** [review]
other than a minor nit...
/lgtm

File: pkg/network/ovn_kubernetes.go
Comment (andreaskaris) [reply pkg/network/ovn_kubernetes.go]: @flavio-fernandes Done, thanks!

File: pkg/network/render.go
Comment (danwinship) [inline pkg/network/render.go:342]: how about "cannot add additional ClusterNetwork values of original IP family when migrating to dual stack" ?

File: pkg/network/render_test.go
Comment (danwinship) [inline pkg/network/render_test.go:99]: You should probably start with an ovn-kubernetes config here. I think an sdn config with IPv6 addresses won't even validate anyway

File: pkg/network/ovn_kubernetes.go
Comment (danwinship) [inline pkg/network/ovn_kubernetes.go:271]: We don't want to just log an error. We want to block migration in this case.

If `isNetworkChangeSafe` doesn't currently have access to the platform type, it can just look it up itself, the same way the bootstrap code does.

**Review by danwinship — CHANGES_REQUESTED** [review]

File: pkg/network/ovn_kubernetes.go
Comment (andreaskaris) [reply pkg/network/ovn_kubernetes.go:271]: Thanks.

~~~
	// obtain the current IP family mode. Only allow configuration of / switch to DualStack on Baremetal or None type platforms.
	ipFamilyMode := names.IPFamilySingleStack
	if len(conf.ServiceNetwork) == 2 {
		if bootstrapResult.Infra.PlatformType == configv1.BareMetalPlatformType || bootstrapResult.Infra.PlatformType == configv1.NonePlatformType {
			ipFamilyMode = names.IPFamilyDualStack
		} else {
			klog.Errorf("Unsupported configuration. DualStack deployments are allowed only for the BareMetal Platform type or the None Platform type. Enforcing SingleStack.")
		}
	}
~~~

The `ipFamilyMode` will remain at `names.IPFamilySingleStack` and is only set to `bootstrapResult.Infra.PlatformType` if either the `BareMetalPlatformType` or `NonePlatformType` is detected. No change in `ipFamilyMode` will be triggered with that code.

If I read the code correctly,  then the chain of events used to be:
len(conf.ServiceNetwork) == 2 --> ipFamilyMode = names.IPFamilyDualStack -> old ipFamilyMode != ipFamilyMode -> trigger migration

The new chain of events with the proposed change would be:
len(conf.ServiceNetwork) == 2 --> ipFamilyMode = names.IPFamilyDualStack -> bootstrapResult.Infra.PlatformType not within required values -> ipFamilyMode = names.IPFamilySingleStack -> old ipFamilyMode == ipFamilyMode -> do not trigger migration.

Is there a more explicit way to block migration?

File: pkg/network/render_test.go
Comment (andreaskaris) [reply pkg/network/render_test.go]: I addressed this and I reordered the test for better readability. I also fixed a typo in one of the error messages:
~~~
-               return []error{errors.Errorf("cannot change migration network type after migration is start")}
+               return []error{errors.Errorf("cannot change migration network type after migration has started")}
~~~

File: pkg/network/ovn_kubernetes.go
Comment (andreaskaris) [reply pkg/network/ovn_kubernetes.go:271]: O.k. after looking at the code again, I'll simply throw an error. I'm currently on that.

Comment (andreaskaris) [reply pkg/network/ovn_kubernetes.go]: With regards to:

>> If isNetworkChangeSafe doesn't currently have access to the platform type, it can just look it up itself, the same way the bootstrap code does.

ReconcileOperConfig.Reconcile(...) runs network.Bootstrap(...) -> boostrapOVN(...) -> platform.BootstrapInfra(...) which determines the Platform type and then immediately after that it runs network.Render(...) ->  renderDefaultNetwork(...) -> renderOVNKubernetes(...). Also, both renderOVNKubernetes and renderDefaultNetwork both use bootstrapResult.Infra.PlatformType at an earlier point in time. Wouldn't another lookup here be redundant?


Before I query the platform type here:
~~~
  67 func renderOVNKubernetes(conf *operv1.NetworkSpec, bootstrapResult *bootstrap.BootstrapResult, manifestDir strin     g) ([]*uns.Unstructured, error) {

(...)
 265         // obtain the current IP family mode. Only allow configuration of / switch to DualStack on Baremetal or      None type platforms.
 266         ipFamilyMode := names.IPFamilySingleStack
 267         if len(conf.ServiceNetwork) == 2 {
 268                 if bootstrapResult.Infra.PlatformType == configv1.BareMetalPlatformType || bootstrapResult.Infra     .PlatformType == configv1.NonePlatformType {
 269                         ipFamilyMode = names.IPFamilyDualStack
 270                 } else {
 271                         return nil, fmt.Errorf("Unsupported configuration. DualStack deployments are allowed onl     y for the BareMetal Platform type or the None Platform type. Halting render operation.")
 272                 }
 273         }
(...)
~~~

We already rely on this to be set, in the same method, here:
~~~
  67 func renderOVNKubernetes(conf *operv1.NetworkSpec, bootstrapResult *bootstrap.BootstrapResult, manifestDir strin     g) ([]*uns.Unstructured, error) {

(...)
 148         if bootstrapResult.Infra.PlatformType == configv1.AzurePlatformType {
 149                 data.Data["OVNPlatformAzure"] = true
 150         } else {
 151                 data.Data["OVNPlatformAzure"] = false
 152         }
(...)
~~~

renderOVNKubernetes is called by:
~~~
460 // renderDefaultNetwork generates the manifests corresponding to the requested
461 // default network
462 func renderDefaultNetwork(conf *operv1.NetworkSpec, bootstrapResult *bootstrap.BootstrapResult, manifestDir strin    g) ([]*uns.Unstructured, error) {
463         dn := conf.DefaultNetwork
464         if errs := validateDefaultNetwork(conf); len(errs) > 0 {
465                 return nil, errors.Errorf("invalid Default Network configuration: %v", errs)
466         }
467 
468         switch dn.Type {
469         case operv1.NetworkTypeOpenShiftSDN:
470                 return renderOpenShiftSDN(conf, bootstrapResult, manifestDir)
471         case operv1.NetworkTypeOVNKubernetes:
472                 return renderOVNKubernetes(conf, bootstrapResult, manifestDir)
473         case operv1.NetworkTypeKuryr:
474                 return renderKuryr(conf, bootstrapResult, manifestDir)
475         default:
476                 log.Printf("NOTICE: Unknown network type %s, ignoring", dn.Type)
477                 return nil, nil
478         }
479 }
~~~

And that in turn by:
~~~
(...)
 23 func Render(conf *operv1.NetworkSpec, bootstrapResult *bootstrap.BootstrapResult, manifestDir string) ([]*uns.Uns    tructured, error) {
 61         // render default network
 62         o, err = renderDefaultNetwork(conf, bootstrapResult, manifestDir)
(...)
~~~

Which also queries the PlatformType before:
~~~
(...)
 27         if bootstrapResult.Infra.PlatformType == v1.AWSPlatformType ||
 28                 bootstrapResult.Infra.PlatformType == v1.AzurePlatformType ||
 29                 bootstrapResult.Infra.PlatformType == v1.GCPPlatformType {
 30                 // render cloud network config controller **before** the network plugin.
 31                 // the network plugin is dependent upon having the cloud network CRD
 32                 // defined as to initialize its watcher, otherwise it will error and crash
 33                 o, err := renderCloudNetworkConfigController(conf, bootstrapResult.Infra, manifestDir)
 34                 if err != nil {
 35                         return nil, err
 36                 }
 37                 objs = append(objs, o...)
 38         }
(...)
~~~

Render is used in:
~~~
(...)
117 func (r *ReconcileOperConfig) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error)     {
(...)
201         // Bootstrap any resources
202         bootstrapResult, err := network.Bootstrap(newOperConfig, r.client)
(...)
219         // Generate the objects
220         objs, err := network.Render(&operConfig.Spec, bootstrapResult, ManifestPath)
(...)
~~~

~~~
 11 // Bootstrap creates resources required by SDN on the cloud.
 12 func Bootstrap(conf *operv1.Network, client client.Client) (*bootstrap.BootstrapResult, error) {
 13         switch conf.Spec.DefaultNetwork.Type {
 14         case operv1.NetworkTypeKuryr:
 15                 return openstack.BootstrapKuryr(&conf.Spec, client)
 16         case operv1.NetworkTypeOpenShiftSDN:
 17                 return bootstrapSDN(conf, client)
 18         case operv1.NetworkTypeOVNKubernetes:
 19                 return bootstrapOVN(conf, client)
 20         }
 21 
 22         return &bootstrap.BootstrapResult{}, nil
 23 }
~~~

~~~
 577 func bootstrapOVN(conf *operv1.Network, kubeClient client.Client) (*bootstrap.BootstrapResult, error) {
(...)
 700         infraRes, err := platform.BootstrapInfra(kubeClient)
 701         if err != nil {
 702                 return nil, err
 703         }
 704         
 705         res := bootstrap.BootstrapResult{
 706                 Infra: *infraRes,
 707                 OVN: bootstrap.OVNBootstrapResult{
 708                         MasterIPs:               ovnMasterIPs,
 709                         ClusterInitiator:        clusterInitiator,
 710                         ExistingMasterDaemonset: masterDS,
 711                         ExistingNodeDaemonset:   nodeDS,
 712                         OVNKubernetesConfig:     ovnConfigResult,
 713                         PrePullerDaemonset:      prePullerDS,
 714                         FlowsConfig:             bootstrapFlowsConfig(kubeClient),
 715                 },
 716         }
 717         return &res, nil
(...)
~~~

~~~
 13 func BootstrapInfra(kubeClient client.Client) (*bootstrap.InfraBootstrapResult, error) {
 14         var platformType configv1.PlatformType
 15         var platformRegion string 
 16 
 17         infraConfig := &configv1.Infrastructure{}
 18         if err := kubeClient.Get(context.TODO(), types.NamespacedName{Name: "cluster"}, infraConfig); err != nil     {
 19                 return nil, fmt.Errorf("failed to get infrastructure 'cluster': %v", err)
 20         }
 21 
 22         if infraConfig.Status.PlatformStatus.Type != "" {
 23                 platformType = infraConfig.Status.PlatformStatus.Type
 24         }
 25         if platformType == configv1.AWSPlatformType {
 26                 platformRegion = infraConfig.Status.PlatformStatus.AWS.Region
 27         } else if platformType == configv1.GCPPlatformType {
 28                 platformRegion = infraConfig.Status.PlatformStatus.GCP.Region
 29         }
 30 
 31         res := &bootstrap.InfraBootstrapResult{
 32                 PlatformType:         platformType,
 33                 PlatformRegion:       platformRegion,
 34                 ExternalControlPlane: infraConfig.Status.ControlPlaneTopology == configv1.ExternalTopologyMode,
 35         }
 36         return res, nil
 37 }
~~~~

Comment (andreaskaris) [reply pkg/network/ovn_kubernetes.go]: After throwing the error, in a live cluster, this now looks like this:
~~~
[akaris@linux 2032559]$ oc patch network.config.openshift.io cluster \
  --type='json' --patch-file patch.yaml 
network.config.openshift.io/cluster patched
[akaris@linux 2032559]$ oc get co | grep network
network                                    4.10.0-0.nightly-2021-12-23-153012   True        False         True       99m     Internal error while rendering operator configuration: Unsupported configuration. DualStack deployments are allowed only for the BareMetal Platform type or the None Platform type. Halting render operation.
~~~

And the cluster logs show:
~~~
(...)
I0106 18:01:49.966009  789421 log.go:184] Reconciling Network.operator.openshift.io cluster
I0106 18:01:49.968292  789421 ovn_kubernetes.go:360] Did not find dpu-mode-config
I0106 18:01:49.969142  789421 log.go:184] Starting render phase
I0106 18:01:49.993319  789421 log.go:184] Failed to render: Unsupported configuration. DualStack deployments are allowed only for the BareMetal Platform type or the None Platform type. Halting render operation.
I0106 18:01:59.855553  789421 log.go:184] Reconciling pki.network.operator.openshift.io openshift-ovn-kubernetes/ovn
I0106 18:01:59.856577  789421 log.go:184] successful reconciliation
I0106 18:02:00.060766  789421 log.go:184] Reconciling pki.network.operator.openshift.io openshift-ovn-kubernetes/signer
I0106 18:02:00.061700  789421 log.go:184] successful reconciliation
I0106 18:02:03.897789  789421 log.go:184] Reconciling update to IngressController openshift-ingress-operator/default
I0106 18:02:12.402369  789421 log.go:184] Reconciling update to openshift-cloud-network-config-controller/cloud-network-config-controller
I0106 18:02:17.211330  789421 log.go:184] Reconciling update to openshift-multus/multus
I0106 18:02:17.652275  789421 log.go:184] Reconciling update to openshift-multus/multus-additional-cni-plugins
I0106 18:02:29.710289  789421 log.go:184] Reconciling update to openshift-ovn-kubernetes/ovnkube-node
I0106 18:03:11.914971  789421 log.go:184] Reconciling Network.operator.openshift.io cluster
I0106 18:03:11.917321  789421 ovn_kubernetes.go:360] Did not find dpu-mode-config
I0106 18:03:11.918121  789421 log.go:184] Starting render phase
I0106 18:03:11.977762  789421 log.go:184] Failed to render: Unsupported configuration. DualStack deployments are allowed only for the BareMetal Platform type or the None Platform type. Halting render operation.
~~~

When updating only the clusternetwork, I get:
~~~
[akaris@linux 2032559]$ oc get co | grep network
network                                    4.10.0-0.nightly-2021-12-23-153012   True        False         True       102m    The cluster configuration is invalid (spec.clusterNetwork and spec.serviceNetwork must either both be IPv4-only, both be IPv6-only, or both be dual-stack). Use 'oc edit network.config.openshift.io cluster' to fix.
~~~

So, that case is covered, too.

Comment (danwinship) [reply pkg/network/ovn_kubernetes.go]: > After throwing the error, in a live cluster, this now looks like this:
> 
>     [akaris@linux 2032559]$ oc get co | grep network
>     network                                    4.10.0-0.nightly-2021-12-23-153012   True        False         True       99m     Internal error while rendering operator configuration: Unsupported configuration. DualStack deployments are allowed only for the BareMetal Platform type or the None Platform type. Halting render operation.

> When updating only the clusternetwork, I get:
> 
>     [akaris@linux 2032559]$ oc get co | grep network
>     network                                    4.10.0-0.nightly-2021-12-23-153012   True        False         True       102m    The cluster configuration is invalid (spec.clusterNetwork and spec.serviceNetwork must either both be IPv4-only, both be IPv6-only, or both be dual-stack). Use 'oc edit network.config.openshift.io cluster' to fix.

The former case should get the same error as the latter; it should fail _before_ it starts trying to render the configuration, and it should give the hint about "use oc edit" in the error message.

Which is to say, you need to fail from `isNetworkChangeSafe`, not from `renderOVNKubernetes`.

Comment (andreaskaris) [reply pkg/network/ovn_kubernetes.go]: I pushed this now to isNetworkChange safe. However, the error messages for both Validate and for IsChangeSafe insinuate to Use 'oc edit network.operator.openshift.io cluster' to undo the change:
~~~
        // Validate the configuration
        if err := network.Validate(&operConfig.Spec); err != nil {
                log.Printf("Failed to validate Network.operator.openshift.io.Spec: %v", err)
                r.status.SetDegraded(statusmanager.OperatorConfig, "InvalidOperatorConfig",
                        fmt.Sprintf("The operator configuration is invalid (%v). Use 'oc edit network.operator.openshift.io cluster' to fix.", err))
                return reconcile.Result{}, err
        }

        // Retrieve the previously applied operator configuration
        prev, err := GetAppliedConfiguration(ctx, r.client, operConfig.ObjectMeta.Name)
        if err != nil {
                log.Printf("Failed to retrieve previously applied configuration: %v", err)
                // FIXME: operator status?
                return reconcile.Result{}, err
        }
        // up-convert Prev by filling defaults
        if prev != nil {
                network.FillDefaults(prev, prev)
        }

        // Fill all defaults explicitly
        network.FillDefaults(&operConfig.Spec, prev)

        // Compare against previous applied configuration to see if this change
        // is safe.
        if prev != nil {
                // We may need to fill defaults here -- sort of as a poor-man's
                // upconversion scheme -- if we add additional fields to the config.
                err = network.IsChangeSafe(prev, &operConfig.Spec, r.client)
                if err != nil {
                        log.Printf("Not applying unsafe change: %v", err)
                        r.status.SetDegraded(statusmanager.OperatorConfig, "InvalidOperatorConfig",
                                fmt.Sprintf("Not applying unsafe configuration change: %v. Use 'oc edit network.operator.openshift.io cluster' to undo the change.", err))
                        return reconcile.Result{}, err
                }
        }
~~~

That's the case both for what I implemented now:
~~~
spec:
  clusterNetwork:
  - cidr: 10.128.0.0/14
    hostPrefix: 23
  - cidr: fd01::/48
    hostPrefix: 64
  externalIP:
    policy: {}
  networkType: OVNKubernetes
  serviceNetwork:
  - 172.30.0.0/16
  - fd02::/112
~~~

~~~
[akaris@linux cluster-network-operator (BZ2032559)]$ oc get co | grep network
network                                    4.10.0-0.nightly-2022-01-10-014106   True        True          True       81m     Not applying unsafe configuration change: invalid configuration: [DualStack deployments are allowed only for the BareMetal Platform type or the None Platform type]. Use 'oc edit network.operator.openshift.io cluster' to undo the change.
~~~

But also for some existing changes, whereas for others not:
~~~
spec:
  clusterNetwork:
  - cidr: 10.128.0.0/14
    hostPrefix: 23
  externalIP:
    policy: {}
  networkType: OVNKubernetes
  serviceNetwork:
  - 172.30.0.0/16
  - fd02::/112
~~~

~~~
[akaris@linux cluster-network-operator (BZ2032559)]$ oc get co | grep network
network                                    4.10.0-0.nightly-2022-01-10-014106   True        True          True       82m     The cluster configuration is invalid (spec.clusterNetwork and spec.serviceNetwork must either both be IPv4-only, both be IPv6-only, or both be dual-stack). Use 'oc edit network.config.openshift.io cluster' to fix
~~~

Not for this one:
~~~
spec:
  clusterNetwork:
  - cidr: 10.128.0.0/14
    hostPrefix: 23
  - cidr: fd01::/48
    hostPrefix: 64
  externalIP:
    policy: {}
  networkType: OVNKubernetes
  serviceNetwork:
  - 172.30.0.0/16
~~~

~~~
[akaris@linux cluster-network-operator (BZ2032559)]$ oc get co | grep network
network                                    4.10.0-0.nightly-2022-01-10-014106   True        True          True       84m     The cluster configuration is invalid (spec.clusterNetwork and spec.serviceNetwork must either both be IPv4-only, both be IPv6-only, or both be dual-stack). Use 'oc edit network.config.openshift.io cluster' to fix.
~~~

But for this one:
~~~
spec:
  clusterNetwork:
  - cidr: 10.128.0.0/14
    hostPrefix: 23
  externalIP:
    policy: {}
  networkType: OVNKubernetes
  serviceNetwork:
  - 172.31.0.0/16
~~~

~~~
[akaris@linux cluster-network-operator (BZ2032559)]$ oc get co | grep network
network                                    4.10.0-0.nightly-2022-01-10-014106   True        True          True       85m     Not applying unsafe configuration change: invalid configuration: [unsupported change to ServiceNetwork]. Use 'oc edit network.operator.openshift.io cluster' to undo the change.
~~~

In both cases where verifications fail, when I edit network.operator.openshift.io cluster, it is immediately reset to whatever network.config.openshift.io cluster dictates.

**Comment by andreaskaris** [comment]
@danwinship Is this better, now? And thanks for the review and help.

File: pkg/network/ovn_kubernetes.go
Comment (danwinship) [reply pkg/network/ovn_kubernetes.go]: Yeah, if they make an illegal change to the `config` object, the error needs to tell them to fix it on the `config` object. If they make an illegal change to the `operator` object, the error needs to tell them to fix it on that object.

Comment (andreaskaris) [reply pkg/network/ovn_kubernetes.go]: Alright. After the latest addition to the `ValidateClusterConfig`, on top of the existing verification for `network.operator.openshift.io`, I get:
~~~
[akaris@linux 2032559]$ oc get co | grep network
network                                    4.10.0-0.nightly-2022-01-18-044014   True        True          True       8h      The cluster configuration is invalid (DualStack deployments are allowed only for the BareMetal Platform type or the None Platform type). Use 'oc edit network.config.openshift.io cluster' to fix.
~~~

**Comment by andreaskaris** [comment]
@danwinship I addressed the latest concern, what do you think of it, now?

**Comment by ardaguclu** [comment]
metal-ipi jobs are failing due to https://github.com/openshift/release/pull/25644.

Old cluster_profile is removed and tests will fail until https://github.com/openshift/release/pull/25644 is merged.

**Comment by danwinship** [comment]
/hold
to stop retests
given that this PR affects behavior on metal, I don't think it makes sense to override the metal tests before merging. Feel free to `/hold cancel` once the metal tests are working again

**Pull request merged**
